src/frameworks/compose/discover/runtime/node.ts

```diff
@@ -0,0 +1,252 @@
+    const nodeVersion = version.node;
+    // Splits version number `>=18..0.5` to `>=` and `18.0.5`
+    const versionPattern = /^([>=<^~]+)?(\d+\.\d+\.\d+)$/;
+    const versionMatch = versionPattern.exec(nodeVersion);
+    if (!versionMatch) {
+      return NODE_LATEST_BASE_IMAGE;
+    }
+    const operator = versionMatch[1];
+    const versionNumber = versionMatch[2];
+    const majorVersion = parseInt(versionNumber.split(".")[0]);
+    if ((!operator || operator === "^" || operator === "~") && majorVersion < 18) {
+      throw new FirebaseError(
+        "Unsupported node version number, only versions >= 18 are supported."
+      );
+    }
+
+    return NODE_LATEST_BASE_IMAGE;
+  }
+
+  async getPackageManager(fs: FileSystem): Promise<string> {
+    if (await fs.exists(YARN_LOCK)) {
+      return YARN;
+    }
+
+    return NPM;
+  }
+
+  async getDependenciesForNPM(
+    fs: FileSystem,
+    packageJSON: PackageJSON
+  ): Promise<Record<string, string>> {
+    const directDependencies = { ...packageJSON.dependencies, ...packageJSON.devDependencies };
+    let transitiveDependencies = {};
+
+    const packageLockJSONRaw = await readOrNull(fs, PACKAGE_LOCK_JSON);
+    if (!packageLockJSONRaw) {
+      return directDependencies;
+    }
+    const packageLockJSON = JSON.parse(packageLockJSONRaw);
+    const directDependencyNames = Object.keys(directDependencies).map((x) =>
+      join("node_modules", x)
+    );
+
+    directDependencyNames.forEach((directDepName) => {
+      const transitiveDeps = packageLockJSON.packages[directDepName].dependencies;
+      transitiveDependencies = { ...transitiveDependencies, ...transitiveDeps };
+    });
+
+    return { ...directDependencies, ...transitiveDependencies };
+  }
+
+  async getDependencies(
+    fs: FileSystem,
+    packageJSON: PackageJSON,
+    packageManager: string
+  ): Promise<Record<string, string>> {
+    try {
+      let dependencies = {};
+      if (packageManager === NPM) {
+        dependencies = await this.getDependenciesForNPM(fs, packageJSON);
+      }
+
+      return dependencies;
+    } catch (error: any) {
+      logger.error("Error while reading dependencies for the project: ", error);
+      throw error;
+    }
+  }
+
+  packageManagerInstallCommand(packageManager: string): string | undefined {
+    const packages: string[] = [];
+    if (packageManager === "yarn") {
+      packages.push("yarn");
+    }
+    if (!packages.length) {
+      return undefined;
+    }
+
+    return `npm install --global ${packages.join(" ")}`;
+  }
+
+  installCommand(packageManager: string): string | undefined {
+    if (packageManager === "npm") {
+      return "npm ci";
+    } else if (packageManager === "yarn") {
+      return "yarn install";
+    }
+
+    return undefined;
+  }
+
+  detectedCommands(
+    packageManager: string,
+    scripts: Record<string, string> | undefined,
+    matchedFramework: FrameworkSpec | null
+  ): LifecycleCommands {
+    const commands: LifecycleCommands = {
```

Comment:
nit* probably better to just return directly instead of creating a intermediate variable?

```
return { build: ... }
```